### ISSUE TITLE
pytest and local pytest plugin running in two separate processes, talking over http

### ISSUE THREAD

**Issue by alanwilter** [pytest-dev/pytest-cov#425]
This is first general approach, I can go to details, but that would require some time since I need to filter out some sensitive data/code. But I summarise the best I can for the moment.

Essentially, we have a local pytest plugin api in `tests/fix_api.py` (snippet):
```python
"""
Pytest plugin to interact with the api at http level
"""

import json
from urllib.parse import urljoin

import pytest
import requests

def pytest_addoption(parser):


@pytest.fixture(scope="session")
def api_url(request):

@pytest.fixture(scope="function")
def api(api_url):

# Users readily available in the test suite
users = {
    "demo": "demo123",
}

class Api:
    def __init__(self, url=None):

    def get(self, user, url, data=None, headers=None):
        return self._request("get", user, url, data=data, headers=headers)

    def put(self, user, url, data=None, headers=None):

    def post(self, user, url, data=None, headers=None):

    def patch(self, user, url, data=None, headers=None):

    def delete(self, user, url, data=None, headers=None):

    def _request(self, method_name, user, url, data=None, headers=None):

    def login(self, user, password=None):

    def logout(self, user):
```
And our `conftest.py`:
```python
import pytest

pytest_plugins = ("tests.fix_api",)
```
Then, a test example `test_auth.py`:
```python
def test_login_logout(api):
    resp = api.get(None, "/is_logged_in")
    assert resp.status_code == 401
    assert resp.json() == {"error": "Unauthenticated"}

    resp = api.post(None, "/login", {"user": "demo", "password": "demo1234"})
    assert resp.status_code == 401
    assert resp.json() == {"error": "Invalid Credentials. Please try again."}

    resp = api.get(None, "/is_logged_in")
    assert resp.status_code == 401

    resp = api.post(None, "/login", {"user": "demo", "password": "demo123"})
    assert resp.status_code == 200
    assert resp.json() == {"success": "Authenticated", "username": "demo"}

    resp = api.get(None, "/is_logged_in")
    assert resp.status_code == 200
    assert resp.json() == {"username": "demo"}

    resp = api.post(None, "/logout")
    assert resp.status_code == 200
    assert resp.json() == {"success": "logged out"}

    resp = api.get(None, "/is_logged_in")
    assert resp.status_code == 401
```
Running `pytest tests/test_auth.py --cov --cov-report term-missing` and I got:
```
views/auth.py                    45     26    42%   20-24, 30-39, 48-55, 63-65, 71
```
Basically, none of the functions `def login()`, `def logout()` and `def is_logged_in()` is being reported as covered.

Since pytest and local pytest plugin are running in two separate processes and talking over http, I'm wondering how could I make, if possible, coverage to work here.

**Comment by ionelmc** [comment]
pytest-cov does in fac support suprocesses. Perhaps there's something weird about how your http server is started. Need details or a reproducer to know for sure.

**Comment by alanwilter** [comment]
@ionelmc Please, see this example I created that illustrates my case: https://github.com/alanwilter/flask-pytest-example

**Comment by ionelmc** [comment]
You need to make sure the subprocess ends cleanly before your test suite completes. See https://pytest-cov.readthedocs.io/en/latest/subprocess-support.html

**Comment by ionelmc** [comment]
Eg, you could do something like this: https://github.com/ionelmc/python-redis-lock/blob/master/tests/conftest.py or make your own subprocess boilerplate that sends sigterm/sigint in the fixture's finalization (after yield).

**Comment by alanwilter** [comment]
Sorry, this is a bit of uncharted waters for me. So, if I understand well, you're telling me I need somehow start `python app.py` and finish it with a pytest session, right?

I will try using subprocess first as I'm more familiar to it (not to redis).

**Comment by ionelmc** [comment]
Well that's one way. Another way is to use whatever flask provides for testing your views in-process.

**Comment by alanwilter** [comment]
In the end, my pytest + cov is running against a docker-compose with db and app running. In these context what would be your advice?

**Comment by ionelmc** [comment]
How about take a look at https://flask.palletsprojects.com/en/1.1.x/testing/

**Comment by alanwilter** [comment]
I've stated with that and that works, if you see the first test in https://github.com/alanwilter/flask-pytest-example/blob/master/tests/test_with_api.py it's a simplification from Flask testing.

My point is I want to use our plugin and I want to add coverage.

I've tweaked the code again. Now my `conftest.py` https://github.com/alanwilter/flask-pytest-example/blob/master/tests/conftest.py has a `@pytest.yield_fixture` and it does start the service (so I don't need `python app.py` in another terminal), yet, running:

```bash
pytest --cov=handlers --cov-report term-missing -k via

handlers/routes.py         6      2    67%   8-10
```
and still don't get 100% coverage.

Certainly I'm missing something. Sorry to bother you again, I'm just new to this all new stuff of Flask and Pytest.

**Comment by alanwilter** [comment]
I got it to work with my main application, but since I want only coverage, I'm not using docker, so I have this:

```python
...
@pytest.fixture(scope="session")
def app_server():
    with TestProcess("python", "application.py") as app_server:
        wait_for_strings(app_server.read, 10, "Running")
        print(app_server.read())
        yield app_server
        print("\n>>>>Teardown app_service")
        app_server.close()

@pytest.fixture(scope="function")
def api(api_url, app_server):
    api = Api(api_url)
    yield api
    for user in list(api.logged_in):
        api.logout(user)
```
so now `test_auth.py` reports 100% covered. However, it's just niggle probably, but I got this message at the end:
```
tests/test_views.py::test_create_user PASSED
>>>>Teardown app_service
2020-09-03 18:31:55,457-ERROR-process_tests::process_tests|186:: TestProcess(pid=79236, is_alive=False) failed to cleanup buffers: ValueError('I/O operation on closed file.')
Traceback (most recent call last):
  File "/usr/local/Caskroom/miniconda/base/lib/python3.7/site-packages/process_tests.py", line 174, in __exit__
    data, _ = self.proc.communicate()
  File "/usr/local/Caskroom/miniconda/base/lib/python3.7/subprocess.py", line 951, in communicate
    stdout = self.stdout.read()
ValueError: I/O operation on closed file.
```
Just wondering what `process_tests:TestProcess` is really trying to do/say.

**Comment by ionelmc** [comment]
@alanwilter I would like to see details, but in the right issue tracker ;) so we can close this.

**Comment by alanwilter** [comment]
You're absolutely right, I'll be fix it, thanks!
